# 📌 Grocery Customer Churn (Abandono de Clientes de Supermercado)

## Acerca del conjunto de datos

Este conjunto de datos contiene información detallada sobre las transacciones, el comportamiento y los datos demográficos de los clientes de una tienda de comestibles. Incluye información de los clientes, datos transaccionales y métricas de comportamiento, lo que lo hace ideal para crear modelos predictivos para la pérdida de clientes y el análisis del valor de vida del cliente (CLV).

La columna objetivo para la predicción de abandono es la tasa de abandono, que indica si un cliente se ha dado de baja (1) o ha permanecido activo (0). El conjunto de datos está diseñado para crear modelos y predecir la retención de clientes, analizar su comportamiento y pronosticar su valor de vida.

Características principales:

* Información del cliente: incluye detalles demográficos como edad, sexo, nivel de ingresos, estado civil, nivel educativo y ocupación.

* Datos transaccionales: contiene información sobre cada transacción, como la fecha de la transacción, la cantidad, el precio, la categoría del producto y el método de pago.

* Métricas de comportamiento del cliente: incluye características como el valor de compra promedio, la frecuencia de compra y la fecha de la última compra.

* Datos promocionales: incluye detalles sobre las promociones, incluido el tipo, la eficacia y el público objetivo.

* Churn: La columna de destino indica si el cliente se ha dado de baja (1) o ha permanecido activo (0).

* Tamaño del conjunto de datos: este conjunto de datos contiene 357.590 filas y 26 atributos (incluida la columna de abandono).

Este estudio incluye las tareas del Data Engineer:

* Obtención y tratamiento de datos a través de una API externa.

* Implementación del proceso ETL (Extracción, Transformación y Carga) para la depuración y organización de la información.

* Desarrollo de visualizaciones clave para detectar patrones y comportamientos relevantes.

* Exploración de los datos (EDA técnico) y elaboración de un reporte con hallazgos significativos, para posterior EDA profundo.

##  Extracción

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv("../data/Grocery_Customer_Churn_Data_Augmented.csv")


## Inspección general del dataset

In [2]:
# Configuraciones de pandas

pd.set_option("display.max_columns", None)
pd.set_option("display.float_format", "{:.2f}".format)


In [3]:
# Mostrar las filas del DataFrame

df.head()
df.tail()



customer_id  age gender income_bracket loyalty_program  \
357552  C_SYNTH_7499   31  Other            Low             Yes   
357553  C_SYNTH_7499   31  Other            Low             Yes   
357554  C_SYNTH_7499   31  Other            Low             Yes   
357555  C_SYNTH_7499   31  Other            Low             Yes   
357556  C_SYNTH_7499   31  Other            Low             Yes   

        membership_years marital_status  number_of_children education_level  \
357552                 9       Divorced                   1     High School   
357553                 9       Divorced                   1     High School   
357554                 9       Divorced                   1     High School   
357555                 9       Divorced                   1     High School   
357556                 9       Divorced                   1     High School   

       occupation    transaction_id transaction_date product_category  \
357552   Employed  T_SYNTH_48fc758a       2024-07-17             Home   
357553   Employed  T_SYNTH_12e530a8       2024-06-08             Home   
357554   Employed  T_SYNTH_2b99d858       2024-07-05             Toys   
357555   Employed  T_SYNTH_90f891dd       2024-07-04         Clothing   
357556   Employed  T_SYNTH_5b37b72f       2024-07-16           Beauty   

        quantity  unit_price  avg_purchase_value purchase_frequency  \
357552         3       45.11               25.58                Low   
357553         1       26.58               25.58                Low   
357554         2       11.32               25.58                Low   
357555         2       53.25               25.58                Low   
357556         2       38.74               25.58                Low   

       last_purchase_date  avg_discount_used  online_purchases  \
357552         2024-12-30               0.09                 2   
357553         2024-12-30               0.09                 2   
357554         2024-12-30               0.09                 2   
357555         2024-12-30               0.09                 2   
357556         2024-12-30               0.09                 2   

        in_store_purchases  total_sales  total_transactions  \
357552                   0        23.28                  43   
357553                   0        23.28                  43   
357554                   0        23.28                  43   
357555                   0        23.28                  43   
357556                   0        23.28                  43   

        total_items_purchased        promotion_type promotion_effectiveness  \
357552                    101                  BOGO                    High   
357553                    101     Seasonal Discount                  Medium   
357554                    101                  BOGO                     Low   
357555                    101  Buy One Get One Free                  Medium   
357556                    101               20% Off                  Medium   

        days_since_last_purchase  churn  
357552                         1      0  
357553                         1      0  
357554                         1      0  
357555                         1      0  
357556                         1      0

In [4]:
# Atributo del DataFrame
df.shape

(357557, 28)

In [5]:
# Obtener información general del DataFrame

df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 357557 entries, 0 to 357556
Data columns (total 28 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   customer_id               357557 non-null  object 
 1   age                       357557 non-null  int64  
 2   gender                    357557 non-null  object 
 3   income_bracket            357557 non-null  object 
 4   loyalty_program           357557 non-null  object 
 5   membership_years          357557 non-null  int64  
 6   marital_status            357557 non-null  object 
 7   number_of_children        357557 non-null  int64  
 8   education_level           357557 non-null  object 
 9   occupation                357557 non-null  object 
 10  transaction_id            357557 non-null  object 
 11  transaction_date          357557 non-null  object 
 12  product_category          357557 non-null  object 
 13  quantity                  357557 non-null  i

**Nota:**

Se observa una mezcla de variables:

* Identificadores (customer_id, transaction_id)

* Variables demográficas del cliente

* Variables transaccionales

* Variables agregadas históricas

Esto indica que el dataset no está normalizado y presenta múltiples niveles de granularidad.

* transaction_date, last_purchase_date → object ❌

* purchase_frequency → debería ser categórica (object está bien, pero validar valores)

* promotion_effectiveness → categórica, no numérica

## Clasificar variables

In [6]:
num_cols = [
    "age", "membership_years", "number_of_children",
    "quantity", "unit_price", "avg_purchase_value",
    "purchase_frequency", "avg_discount_used",
    "online_purchases", "in_store_purchases",
    "total_sales", "total_transactions",
    "total_items_purchased", "promotion_effectiveness",
    "days_since_last_purchase"
]

categorical_cols = [
    "gender", "income_bracket", "loyalty_program",
    "marital_status", "education_level", "occupation",
    "product_category", "promotion_type", "churn"
]

date_cols = [
    "transaction_date", "last_purchase_date"
]

id_cols = [
    "customer_id", "transaction_id"
]


Se realizó una clasificación explícita de columnas por tipo lógico (numéricas, categóricas, fechas e identificadores) con el objetivo de:

* Facilitar validaciones de calidad de datos

* Aplicar transformaciones específicas por tipo

* Mejorar la mantenibilidad del pipeline ETL

## Detección de valores nulos

In [7]:
nulls = df.isna().sum().sort_values(ascending=False)
nulls[nulls > 0]


avg_purchase_value    36355
total_sales           34785
purchase_frequency    19941
promotion_type        14241
dtype: int64

**Nota:**

Se identifican valores faltantes principalmente en variables relacionadas con promociones, frecuencia de compra y métricas agregadas.

Columnas con mayor cantidad de nulos:

* promotion_type: 14241
* avg_purchase_value: 3531
* purchase_frequency: 1779
* total_sales: 3614

Esto sugiere que:

* No todas las transacciones estuvieron asociadas a promociones

* Algunas métricas agregadas pueden no haber sido calculadas para todos los registros

Estos nulos no se consideran errores automáticamente, sino que requieren una decisión de negocio (imputación, categorización explícita o exclusión según el caso).

* promotion_type nulo → no todas las transacciones tuvieron promoción

* purchase_frequency y avg_purchase_value → no estén calculadas para todos los clientes

* total_sales nulo → inconsistencia, posiblemente falte para clientes sin historial consolidado


## Duplicados y unicidad

In [8]:
df["customer_id"].duplicated().sum()
df["transaction_id"].duplicated().sum()

np.int64(10)

In [9]:
df.groupby("customer_id")["transaction_id"].nunique().describe()


count   5000.00
mean      71.51
std        0.50
min       71.00
25%       71.00
50%       72.00
75%       72.00
max       72.00
Name: transaction_id, dtype: float64

In [10]:
df["transaction_id"].nunique()

357547

In [11]:
df["customer_id"].value_counts().head()

customer_id
C_SYNTH_7499    72
C1000           72
C1001           72
C1002           72
C1003           72
Name: count, dtype: int64

**Nota:**

Conclusión estructural:

* Cada transaction_id aparece ~72 veces

* Pero cada customer_id aparece muchas veces

* transaction_id NO es una transacción. Además  transaction_id identifica un CUSTOMER, no un evento




## EDA técnico

### Variables Númericas y Fechas

In [12]:
df[num_cols ].describe()

age  membership_years  number_of_children  quantity  unit_price  \
count 357557.00         357557.00           357557.00 357557.00   357557.00   
mean      44.27              7.68                2.11      2.57       50.15   
std       14.93              4.44                1.42      1.05       20.00   
min       18.00              1.00                0.00      1.00        0.01   
25%       32.00              4.00                1.00      2.00       36.54   
50%       45.00              8.00                2.00      3.00       49.99   
75%       57.00             11.00                3.00      3.00       63.55   
max       70.00             15.00                4.00     12.00      199.75   

       avg_purchase_value  avg_discount_used  online_purchases  \
count           321202.00          357557.00         357557.00   
mean               134.02               2.09              2.62   
std                 85.32               4.14              2.41   
min                 -1.77               0.00              0.00   
25%                 68.94               0.09              0.00   
50%                120.25               0.18              2.00   
75%                178.92               0.28              4.00   
max                909.10              14.96             12.00   

       in_store_purchases  total_sales  total_transactions  \
count           357557.00    322772.00           357557.00   
mean                 1.60       649.39               29.29   
std                  2.44      1253.41               13.56   
min                  0.00      -896.23                5.00   
25%                  0.00        70.98               18.00   
50%                  0.00       130.09               29.00   
75%                  3.00       250.74               42.00   
max                 11.00      6233.02               51.00   

       total_items_purchased  days_since_last_purchase  
count              357557.00                 357557.00  
mean                   57.39                    118.84  
std                    26.41                    252.35  
min                    10.00                   -318.00  
25%                    33.00                      2.00  
50%                    59.00                      7.00  
75%                    79.00                    124.00  
max                   109.00                   1273.00

In [13]:
# Lista de columnas numéricas a evaluar

# Función para detectar outliers por regla de negocio
def detect_anomalies(df, cols):
    anomalies = {}

    for col in cols:
        if col not in df.columns:
            continue

        # Forzar conversión numérica
        series_raw = df[col]
        series = pd.to_numeric(series_raw, errors="coerce")

        n_missing = series.isnull().sum()
        n_negative = (series < 0).sum()
        n_zero = (series == 0).sum()

        min_val = series.min()
        max_val = series.max()
        mean_val = series.mean()
        std_val = series.std()

        if pd.notna(std_val) and std_val > 0:
            lower_limit = mean_val - 3 * std_val
            upper_limit = mean_val + 3 * std_val
            n_outliers = ((series < lower_limit) | (series > upper_limit)).sum()
        else:
            n_outliers = 0

        anomalies[col] = {
            "missing": int(n_missing),
            "negative": int(n_negative),
            "zero": int(n_zero),
            "min": min_val,
            "max": max_val,
            "mean": mean_val,
            "std": std_val,
            "outliers_3sigma": int(n_outliers)
        }

    return pd.DataFrame(anomalies).T


In [14]:
# Evaluación numéricas
num_anomalies = detect_anomalies(df, num_cols)
display(num_anomalies.sort_values(by="outliers_3sigma", ascending=False))

# Fechas: detectar fechas futuras o inconsistentes
today = pd.Timestamp.today()

date_cols = ["transaction_date", "last_purchase_date"]

for col in date_cols:
    if col in df.columns:
        df[col] = pd.to_datetime(df[col], errors="coerce")


for col in date_cols:
    if col in df.columns:
        n_missing = df[col].isnull().sum()
        n_future = (df[col] > today).sum()
        min_date = df[col].min()
        max_date = df[col].max()

        print(
            f"{col}: "
            f"missing={n_missing}, "
            f"future_dates={n_future}, "
            f"min={min_date}, "
            f"max={max_date}"
        )


missing  negative      zero     min     max   mean  \
days_since_last_purchase      0.00    318.00  40678.00 -318.00 1273.00 118.84   
in_store_purchases            0.00      0.00 207105.00    0.00   11.00   1.60   
total_sales               34785.00   1524.00      0.00 -896.23 6233.02 649.39   
online_purchases              0.00      0.00  93848.00    0.00   12.00   2.62   
avg_discount_used             0.00      0.00   2848.00    0.00   14.96   2.09   
quantity                      0.00      0.00      0.00    1.00   12.00   2.57   
avg_purchase_value        36355.00      2.00      0.00   -1.77  909.10 134.02   
unit_price                    0.00      0.00      0.00    0.01  199.75  50.15   
age                           0.00      0.00      0.00   18.00   70.00  44.27   
membership_years              0.00      0.00      0.00    1.00   15.00   7.68   
number_of_children            0.00      0.00  60751.00    0.00    4.00   2.11   
purchase_frequency       357557.00      0.00      0.00     NaN     NaN    NaN   
total_transactions            0.00      0.00      0.00    5.00   51.00  29.29   
total_items_purchased         0.00      0.00      0.00   10.00  109.00  57.39   
promotion_effectiveness  357557.00      0.00      0.00     NaN     NaN    NaN   

                             std  outliers_3sigma  
days_since_last_purchase  252.35         15583.00  
in_store_purchases          2.44         10399.00  
total_sales              1253.41          9926.00  
online_purchases            2.41          6414.00  
avg_discount_used           4.14          5757.00  
quantity                    1.05          2227.00  
avg_purchase_value         85.32          1933.00  
unit_price                 20.00           731.00  
age                        14.93             0.00  
membership_years            4.44             0.00  
number_of_children          1.42             0.00  
purchase_frequency           NaN             0.00  
total_transactions         13.56             0.00  
total_items_purchased      26.41             0.00  
promotion_effectiveness      NaN             0.00

transaction_date: missing=0, future_dates=0, min=2022-07-05 00:00:00, max=2024-12-31 00:00:00
last_purchase_date: missing=0, future_dates=0, min=2022-07-05 00:00:00, max=2024-12-31 00:00:00


**Nota:**

1️⃣ Columnas numéricas

* age, membership_years, number_of_children, quantity, unit_price

Sin valores faltantes

Sin valores negativos

Sin outliers según criterio 3σ
✅ Columnas saneadas y confiables

* avg_purchase_value

3.531 valores nulos → ya identificado previamente

2 valores negativos

Mínimo = -1.77 → valores pequeños, probablemente errores de transacción o ajustes mal calculados
⚠️ Requiere limpieza puntual (filtrar o imputar)

* purchase_frequency

35.844 valores nulos (100%)

➡️ Debe recalcularse

* total_sales

3.614 valores nulos → ya identificado previamente

18 valores negativos

Mínimo = -896.23 → errores claros de cálculo o registros de reverso mal tratados
⚠️ Revisar fórmula

* online_purchases, in_store_purchases
Alta cantidad de ceros

online_purchases: 10.626 ceros

in_store_purchases: 24.812 ceros
ℹ️ Puede ser consistente con el comportamiento del cliente (canales excluyentes)
➡️ Validar contra reglas de negocio

* promotion_effectiveness

35.844 valores nulos (100%)
❌ Variable inutilizable en su estado actual
➡️ Recalcular o eliminar

* days_since_last_purchase

318 valores negativos

Mínimo = -318
❌ Error sistémico confirmado
➡️ El problema no es la fecha en sí, sino el cálculo

2️⃣ Columnas de fecha

* transaction_date y last_purchase_date

No hay nulos, no hay fechas futuras → ✅ consistentes


3️⃣ Conclusiones parciales

Las fechas son confiables, no hay que imputar ni corregir transaction_date ni last_purchase_date.

days_since_last_purchase debe recalcularse para eliminar los negativos.

purchase_frequency y promotion_effectiveness no se pueden usar tal cual, requieren recalculo o imputación.

avg_purchase_value y total_sales tienen algunos valores negativos. Revisar cómo se calcularon (probablemente quantity * unit_price para total_sales).

El resto de variables numéricas están limpias y listas para análisis o derivación de features.

In [15]:
(df["age"] < 0).sum(), (df["age"] > 100).sum()


(np.int64(0), np.int64(0))

In [16]:
(df["unit_price"] < 0).sum()


np.int64(0)

In [17]:
(df["quantity"] <= 0).sum()


np.int64(0)

In [18]:
(df["transaction_date"] > pd.Timestamp.today()).sum()


np.int64(0)

In [19]:
# Detección de anomalías lógicas

anomalies = {}

# Ventas negativas
anomalies["negative_total_sales"] = df[df["total_sales"] < 0].shape[0]

# Cantidades inválidas
anomalies["invalid_quantity"] = df[df["quantity"] <= 0].shape[0]

# Precios inválidos
anomalies["invalid_unit_price"] = df[df["unit_price"] <= 0].shape[0]

# Edades fuera de rango lógico
anomalies["invalid_age"] = df[(df["age"] < 18) | (df["age"] > 100)].shape[0]

# Métricas temporales inconsistentes
if "days_since_last_purchase" in df.columns:
    anomalies["negative_days_since_last_purchase"] = (
        df[df["days_since_last_purchase"] < 0].shape[0]
    )

# Resumen de anomalías
anomaly_df = pd.DataFrame.from_dict(
    anomalies, orient="index", columns=["count"]
)

print("\nResumen de anomalías detectadas:")
display(anomaly_df)


Resumen de anomalías detectadas:


count
negative_total_sales                1524
invalid_quantity                       0
invalid_unit_price                     0
invalid_age                            0
negative_days_since_last_purchase    318

**Nota:**

El análisis de calidad de datos identificó dos tipos principales de anomalías:

* Ventas totales negativas (negative_total_sales):
Se detectaron 18 registros con valores negativos, lo que puede estar asociado a devoluciones, cancelaciones o ajustes contables. Estos casos requieren validación de negocio para decidir si deben mantenerse, corregirse o excluirse.

* Días negativos desde la última compra (negative_days_since_last_purchase):
Se identificaron 318 registros con valores negativos, lo que indica inconsistencias temporales (por ejemplo, fechas de referencia mal definidas o eventos posteriores al “snapshot” de análisis). Esta anomalía sugiere un problema de alineación temporal más que de valores numéricos.

* No se detectaron anomalías en:

1.Cantidades (invalid_quantity)

2.Precios unitarios (invalid_unit_price)

3.Edad de clientes (invalid_age)

### Variables Catégoricas

In [20]:
for col in categorical_cols:
    print("\n", col)
    display(df[col].value_counts(dropna=False).head(10))



 gender


gender
Male      121562
Other     118996
Female    116999
Name: count, dtype: int64


 income_bracket


income_bracket
High      138458
Low       113281
Medium    105818
Name: count, dtype: int64


 loyalty_program


loyalty_program
Yes    187812
No     169745
Name: count, dtype: int64


 marital_status


marital_status
Single      133430
Married     112206
Divorced    111921
Name: count, dtype: int64


 education_level


education_level
Master's       95960
Bachelor's     91770
PhD            91025
High School    78802
Name: count, dtype: int64


 occupation


occupation
Employed         79019
Self-Employed    76141
Unemployed       71424
Retired          70667
Student          60306
Name: count, dtype: int64


 product_category


product_category
Beauty         41890
Sports         41815
Electronics    41725
Clothing       41701
Books          41546
Home           41478
Groceries      35962
Home Goods     35785
Toys           35655
Name: count, dtype: int64


 promotion_type


promotion_type
Discount                74997
BOGO                    74762
Seasonal Discount       64679
Buy One Get One Free    64511
20% Off                 64367
NaN                     14241
Name: count, dtype: int64


 churn


churn
0    265903
1     91654
Name: count, dtype: int64

**Nota:**

1️⃣ Género

Male: 12.116

Other: 12.112

Female: 11.615

✅ Hay una distribución bastante equilibrada entre géneros, con una ligera ventaja de “Male” y “Other” sobre “Female”. Esto es útil, porque no hay un sesgo fuerte hacia un género en los análisis posteriores.

2️⃣ Income Bracket (Nivel de ingreso)

High: 13.769

Low: 11.747

Medium: 10.328

✅ Predomina el nivel High, seguido de Low y Medium. Esto indica que la base de clientes tiene una proporción relativamente alta de ingresos altos, lo que puede influir en patrones de consumo y promociones.

3️⃣ Loyalty Program

Yes: 18.845

No: 16.999

✅ La mayoría de los clientes participa en un programa de fidelidad, pero hay una proporción significativa que no lo hace. Esto es clave para análisis de retención o churn.

4️⃣ Marital Status

Single: 13.182

Divorced: 11.333

Married: 11.329

✅ Hay más clientes solteros, aunque los divorciados y casados están bastante equilibrados. Esto puede impactar en patrones de compra según la etapa de vida.

5️⃣ Education Level

Master’s: 9.826

PhD: 9.108

Bachelor’s: 8.810

High School: 8.100

✅ La mayoría tiene educación avanzada, con predominio de Master’s y PhD. Esto podría correlacionarse con ingresos altos y ciertos hábitos de consumo.

6️⃣ Occupation

Employed: 7.812

Self-Employed: 7.750

Unemployed: 7.239

Retired: 6.880

Student: 6.163

✅ Hay una diversidad de ocupaciones, siendo la mayoría empleados o autónomos. Esto puede afectar frecuencia de compra y preferencias de productos.

7️⃣ Product Category

Electronics: 6.020

Home: 5.969

Books: 5.949

Sports: 5.914

Clothing: 5.856

Beauty: 5.838

Home Goods: 112

Toys: 102

Groceries: 84

✅ Las categorías más frecuentes son Electronics, Home, Books y Sports. Groceries, Toys y Home Goods son muy poco frecuentes, indicando enfoque en bienes duraderos o consumo discrecional.

8️⃣ Promotion Type

NaN: 14.241

Discount: 10.617

BOGO: 10.566

Seasonal Discount: 151

20% Off: 136

Buy One Get One Free: 133

✅ La mayoría de las transacciones no tienen promoción registrada (≈40%), pero cuando hay, predominan Discount y BOGO. Las promociones siguen siendo un factor relevante para analizar su efecto en ventas y churn

9️⃣ Churn

0 (No churn): 35.700

1 (Churn): 144

✅ La tasa de churn es extremadamente baja (<0,5%), lo que indica que la mayoría de los clientes permanecen activos.



Conclusión general

* La base de clientes es bastante equilibrada en género y estado civil, aunque ligeramente inclinada hacia solteros y hombres/other.

* Hay predominio de ingresos altos y educación avanzada, lo que podría afectar patrones de gasto.

* La participación en programas de fidelidad es alta, lo que permite analizar retención y efectividad de promociones.

* Los productos más populares son Electronics, Home, Books y Sports, y la mayoría de las compras incluyen alguna promoción.

* Los datos son en general balanceados, pero hay algunos valores faltantes en promotion_type, avg_purchase_value y otras variables que deberíamos considerar antes de modelar.

## Limpieza numérica y nulos

In [21]:
df_clean = df.copy()

In [22]:
# Columnas binarias que queremos inspeccionar
binary_cols = ["loyalty_program", "churn", "promo_flag"]

for col in binary_cols:
    if col in df_clean.columns:
        print(f"Valores únicos en '{col}':")
        print(df_clean[col].value_counts(dropna=False))
        print("-" * 40)


Valores únicos en 'loyalty_program':
loyalty_program
Yes    187812
No     169745
Name: count, dtype: int64
----------------------------------------
Valores únicos en 'churn':
churn
0    265903
1     91654
Name: count, dtype: int64
----------------------------------------


In [23]:
df_clean = df.copy()

df_clean["gender"] = df_clean["gender"].fillna("unknown")

df_clean["age"] = df_clean["age"].clip(0, 100)
df_clean["age"] = df_clean["age"].fillna(df_clean["age"].median())

df_clean["unit_price"] = df_clean["unit_price"].clip(lower=0)


In [24]:
# Normalizar nombres de columnas
def clean_column_names(df_clean):
    df_clean = df.copy()
    df_clean.columns = (
        df.columns
        .str.strip()                # Quitar espacios al inicio/final
        .str.lower()                # Pasar todo a minúsculas
        .str.replace(" ", "_")      # Reemplazar espacios por _
        .str.replace("-", "_")      # Reemplazar guiones por _
        .str.replace(r"[^\w_]", "", regex=True)  # Quitar caracteres especiales
    )
    return df_clean

## Convertir fechas

In [25]:
# Convertir fechas
date_cols = ["transaction_date", "last_purchase_date"]
for col in date_cols:
    df_clean[col] = pd.to_datetime(df_clean[col], errors="coerce")

today = pd.Timestamp.today().normalize()

# Detectar y reportar fechas inconsistentes
for col in date_cols:
    n_missing = df[col].isnull().sum()
    n_future = (df[col] > today).sum()
    print(f"{col}: missing={n_missing}, future_dates={n_future}")

# Recalcular métrica correctamente
df_clean["days_since_last_purchase"] = (
    today - df_clean["last_purchase_date"]
).dt.days

# Seguridad final
df_clean.loc[
    df_clean["days_since_last_purchase"] < 0,
    "days_since_last_purchase"
] = 0

transaction_date: missing=0, future_dates=0
last_purchase_date: missing=0, future_dates=0


## Normalización de categorías y reducción de cardinalidad

In [26]:
for col in categorical_cols:

    # Reducir categorías raras (<1% frecuencia)
    freq = df_clean[col].value_counts(normalize=True)
    rare_labels = freq[freq < 0.01].index
    df_clean[col] = df_clean[col].replace(rare_labels, "other")

# Promotion type: rellenar NaN
df_clean["promotion_type"] = df_clean["promotion_type"].fillna("no_promotion")


In [27]:
df_clean["loyalty_program"] = df_clean["loyalty_program"].astype(str).str.strip().str.lower()

# Mapear Yes/No a 1/0
df_clean["loyalty_program"] = df_clean["loyalty_program"].map({"yes": 1, "no": 0})


## Métricas derivadas

In [28]:
df_clean["total_sales"] = df_clean["quantity"] * df_clean["unit_price"]
df_clean["avg_purchase_value"] = df_clean.groupby("customer_id")["total_sales"].transform("mean")
df_clean["purchase_frequency"] = df_clean.groupby("customer_id")["transaction_id"].transform("count") / df_clean["membership_years"].replace(0,1)

# Promotion effectiveness
df_clean["promo_flag"] = df_clean["promotion_type"].apply(lambda x: 0 if x=="no_promotion" else 1)
df_clean["promotion_effectiveness"] = df_clean.groupby("customer_id")["promo_flag"].transform("sum") / df_clean.groupby("customer_id")["transaction_id"].transform("count")

# Online vs in-store ratio
df_clean["total_purchases"] = df_clean["online_purchases"] + df_clean["in_store_purchases"]
df_clean["online_ratio"] = df_clean["online_purchases"] / df_clean["total_purchases"].replace(0,1)


In [29]:
# Mapear churn correctamente
df_clean['churn'] = df_clean['churn'].replace({'other': 1})

# Convertir a int
df_clean['churn'] = df_clean['churn'].astype(int)

# Verificar
print(df_clean['churn'].value_counts())

churn
0    265903
1     91654
Name: count, dtype: int64


## Validación post-transformación

In [30]:
for col in num_cols:
    if col in df_clean.columns:
        print(f"{col}: min={df_clean[col].min()}, max={df_clean[col].max()}, mean={df_clean[col].mean():.2f}, nulos={df_clean[col].isnull().sum()}")

age: min=18, max=70, mean=44.27, nulos=0
membership_years: min=1, max=15, mean=7.68, nulos=0
number_of_children: min=0, max=4, mean=2.11, nulos=0
quantity: min=1, max=12, mean=2.57, nulos=0
unit_price: min=0.01, max=199.7514692043265, mean=50.15, nulos=0
avg_purchase_value: min=96.1725, max=190.00896040297204, mean=129.02, nulos=0
purchase_frequency: min=4.733333333333333, max=72.0, mean=17.99, nulos=0
avg_discount_used: min=0.0, max=14.96257824333469, mean=2.09, nulos=0
online_purchases: min=0, max=12, mean=2.62, nulos=0
in_store_purchases: min=0, max=11, mean=1.60, nulos=0
total_sales: min=0.01, max=990.1493097576405, mean=129.02, nulos=0
total_transactions: min=5, max=51, mean=29.29, nulos=0
total_items_purchased: min=10, max=109, mean=57.39, nulos=0
promotion_effectiveness: min=0.4166666666666667, max=1.0, mean=0.96, nulos=0
days_since_last_purchase: min=367, max=1277, mean=450.73, nulos=0


In [31]:
# Dataset final con todas las columnas procesadas
df_final = df_clean
df_final = df_final.reset_index(drop=True)
df_final.to_csv("../data/dataset_final_all_columns.csv", index=False)

## Dataset Data Analyst

In [32]:
# Dataset resumido por cliente

df_agg = df_final.groupby("customer_id").agg({
    # Numéricas: agregaciones típicas
    "age": "first",
    "membership_years": "first",
    "number_of_children": "first",
    "quantity": "mean",
    "unit_price": "mean",
    "avg_purchase_value": "mean",
    "purchase_frequency": "mean",
    "avg_discount_used": "mean",
    "online_purchases": "sum",
    "in_store_purchases": "sum",
    "total_sales": "sum",
    "total_transactions": "sum",
    "total_items_purchased": "sum",
    "promotion_effectiveness": "mean",
    "days_since_last_purchase": "min",

    # Categóricas: valor más frecuente
    "gender": lambda x: x.mode()[0],
    "income_bracket": lambda x: x.mode()[0],
    "marital_status": lambda x: x.mode()[0],
    "education_level": lambda x: x.mode()[0],
    "occupation": lambda x: x.mode()[0],
    "product_category": lambda x: x.mode()[0],
    "promotion_type": lambda x: x.mode()[0],


    # Fechas
    "transaction_date": "max",
    "last_purchase_date": "max",

    # Flags binarios
    "loyalty_program": "max",
    "churn": "max",
    "promo_flag": "max"

}).reset_index()


# Exportar CSV para Data Analyst

df_agg.to_csv("../data/dataset_analyst_by_customer.csv", index=False)


**Documentación para Data analyst:**


* Diccionario de columnas con descripción y tipo.

* Rango y estadística clave de cada variable.

* Notas sobre imputaciones, clipping y derivadas.

## Dataset Data Scientist

### Codificación de variables categóricas: One-hot encoding

In [33]:
categorical_cols = [
    "gender", "income_bracket", "marital_status",
    "education_level", "occupation",
    "product_category", "promotion_type"
]

df_final_encoded = pd.get_dummies(df_agg, columns=categorical_cols, drop_first=False)

In [34]:
# Dataset final para Data Scientist
df_final_encoded.to_csv("../data/customer_dataset_for_ml.csv", index=False)

**Documentación para Data Scientist:**


* Diccionario de columnas codificadas (qué representa cada columna 0/1).

* Notas sobre limpieza y derivadas.

* Estadísticas resumidas de features numéricas.

## Validación final

### 1️⃣ Chequeo rápido de nulos / NaN

In [35]:
def check_nulls(df, name="dataset"):
    print(f"\n📌 Null check → {name}")
    nulls = df.isna().sum()
    nulls = nulls[nulls > 0]

    if nulls.empty:
        print("✅ No hay valores nulos ni NaN")
    else:
        print("❌ Columnas con nulos:")
        print(nulls)


In [36]:
check_nulls(df_final, "df_final")
check_nulls(df_agg, "df_agg")



📌 Null check → df_final
✅ No hay valores nulos ni NaN

📌 Null check → df_agg
✅ No hay valores nulos ni NaN


### 2️⃣Verificar columnas binarias (0/1)

In [37]:
def check_binary_cols(df, binary_cols):
    print("\n📌 Binary columns check")
    for col in binary_cols:
        if col in df.columns:
            invalid = df[~df[col].isin([0,1]) & df[col].notna()]
            if len(invalid) == 0:
                print(f"✅ {col}: OK (solo 0/1)")
            else:
                print(f"❌ {col}: valores inválidos →")
                print(df[col].value_counts())


In [38]:
binary_cols = ["loyalty_program", "churn", "promo_flag" ]
check_binary_cols(df_final, binary_cols)



📌 Binary columns check
✅ loyalty_program: OK (solo 0/1)
✅ churn: OK (solo 0/1)
✅ promo_flag: OK (solo 0/1)


### 3️⃣ Verificación de rangos numéricos

In [39]:
def check_numeric_ranges(df, num_cols):
    print("\n📌 Numeric ranges check")
    for col in num_cols:
        if col in df.columns:
            min_v = df[col].min()
            max_v = df[col].max()
            if min_v < 0:
                print(f"⚠️ {col}: valor negativo detectado (min={min_v})")
            else:
                print(f"✅ {col}: min={min_v}, max={max_v}")


In [40]:
check_numeric_ranges(df_final, num_cols)



📌 Numeric ranges check
✅ age: min=18, max=70
✅ membership_years: min=1, max=15
✅ number_of_children: min=0, max=4
✅ quantity: min=1, max=12
✅ unit_price: min=0.0, max=199.7514692043265
✅ avg_purchase_value: min=96.11718309859154, max=190.00896040297204
✅ purchase_frequency: min=4.733333333333333, max=72.0
✅ avg_discount_used: min=0.0, max=14.96257824333469
✅ online_purchases: min=0, max=12
✅ in_store_purchases: min=0, max=11
✅ total_sales: min=0.0, max=990.1493097576405
✅ total_transactions: min=5, max=51
✅ total_items_purchased: min=10, max=109
✅ promotion_effectiveness: min=0.4166666666666667, max=1.0
✅ days_since_last_purchase: min=418, max=1277


### 4️⃣ Chequeo de fechas (consistencia temporal)

In [41]:
def check_dates(df, date_cols):
    print("\n📌 Date columns check")
    for col in date_cols:
        if col in df.columns:
            if not pd.api.types.is_datetime64_any_dtype(df[col]):
                print(f"❌ {col}: no es datetime")
            else:
                print(f"✅ {col}: OK ({df[col].min()} → {df[col].max()})")


In [42]:
check_dates(df_final, date_cols)



📌 Date columns check
✅ transaction_date: OK (2022-07-05 00:00:00 → 2024-11-10 00:00:00)
✅ last_purchase_date: OK (2022-07-05 00:00:00 → 2024-11-10 00:00:00)


### 5️⃣ Valores categóricos raros (cardinalidad)

In [43]:
def check_categorical_cols(df, cat_cols, max_unique=20):
    print("\n📌 Categorical consistency check")
    for col in cat_cols:
        if col in df.columns:
            uniq = df[col].nunique(dropna=False)
            print(f"🔹 {col}: {uniq} valores únicos")
            if uniq <= max_unique:
                print(df[col].value_counts())


In [44]:
check_categorical_cols(df_agg, categorical_cols)



📌 Categorical consistency check
🔹 gender: 3 valores únicos
gender
Other     1680
Male      1678
Female    1642
Name: count, dtype: int64
🔹 income_bracket: 3 valores únicos
income_bracket
High      1852
Low       1692
Medium    1456
Name: count, dtype: int64
🔹 marital_status: 3 valores únicos
marital_status
Single      1890
Divorced    1614
Married     1496
Name: count, dtype: int64
🔹 education_level: 4 valores únicos
education_level
Master's       1314
PhD            1307
Bachelor's     1238
High School    1141
Name: count, dtype: int64
🔹 occupation: 5 valores únicos
occupation
Self-Employed    1066
Employed         1060
Unemployed       1034
Retired           977
Student           863
Name: count, dtype: int64
🔹 product_category: 9 valores únicos
product_category
Beauty         739
Books          672
Clothing       642
Electronics    624
Home           532
Groceries      501
Sports         488
Home Goods     431
Toys           371
Name: count, dtype: int64
🔹 promotion_type: 6 valores

### 6️⃣ Chequeo de duplicados

In [45]:
def check_duplicates(df, subset=None, name="dataset"):
    print(f"\n📌 Duplicate check → {name}")
    dups = df.duplicated(subset=subset).sum()
    if dups == 0:
        print("✅ No hay duplicados")
    else:
        print(f"❌ {dups} duplicados encontrados")


In [46]:
check_duplicates(df_agg, subset=["customer_id"], name="df_agg")



📌 Duplicate check → df_agg
✅ No hay duplicados


### 7️⃣ Check FINAL tipo “semáforo”

In [47]:
def final_dataset_check(df, name):
    print(f"\n🚦 FINAL CHECK → {name}")
    print("Filas:", df.shape[0])
    print("Columnas:", df.shape[1])
    print("NaN totales:", df.isna().sum().sum())


In [48]:
final_dataset_check(df_final, "Dataset Final (All Columns)")
final_dataset_check(df_agg, "Dataset Analyst by Customer")



🚦 FINAL CHECK → Dataset Final (All Columns)
Filas: 357589
Columnas: 31
NaN totales: 0

🚦 FINAL CHECK → Dataset Analyst by Customer
Filas: 5000
Columnas: 28
NaN totales: 0
